### PR TITLE
Throw from extension methods not supported by WebApplicationBuilder

### LINKED ISSUE #34824: Add ISupportsConfigureWebHost interface to hosting.
## API Proposal

```diff
namespace Microsoft.AspNetCore.Hosting.Infrastructure
{
+    public interface ISupportsStartup
+    {
+        IWebHostBuilder Configure(Action<WebHostBuilderContext, IApplicationBuilder> configure);
+        IWebHostBuilder UseStartup([DynamicallyAccessedMembers(StartupLinkerOptions.Accessibility)] Type startupType);
+        IWebHostBuilder UseStartup<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]TStartup>(Func<WebHostBuilderContext, TStartup> startupFactory);
+    }
+
+    public interface ISupportsConfigureWebHost
+    {
+        IHostBuilder ConfigureWebHost(Action<IWebHostBuilder> configure, Action<WebHostBuilderOptions> configureOptions);
+    }
```
We would implement `ISupportsStartup` in [ConfigureWebHostBuilder](https://github.com/dotnet/aspnetcore/blob/a7ca111caeb3f1a66a65ca802a3b7d18617b2e80/src/DefaultBuilder/src/ConfigureWebHostBuilder.cs) (i.e. `WebApplicationBuilder.WebHost`) and `ISupportsConfigureWebHost` in [ConfigureHostBuilder](https://github.com/dotnet/aspnetcore/blob/a7ca111caeb3f1a66a65ca802a3b7d18617b2e80/src/DefaultBuilder/src/ConfigureHostBuilder.cs) (i.e. `WebApplicationBuilder.Host`) .

## Justifcation

In order to make `builder.WebHost.UseStartup` and `builder.WebHost.Configure` throw instead of no-op in the following:

```c#
var builder = WebApplication.CreateBuilder(args);
builder.WebHost.UseStartup<Startup>();
var app = builder.Build();
```

```c#
var builder = WebApplication.CreateBuilder(args);
builder.WebHost.Configure(app =>
{
    app.Run(async context =>
    {
        await context.Response.WriteAsync("Hello world from configure!");
    });
});
var app = builder.Build();
```

This is because both [UseStartup](https://github.com/dotnet/aspnetcore/blob/a7ca111caeb3f1a66a65ca802a3b7d18617b2e80/src/Hosting/Hosting/src/WebHostBuilderExtensions.cs#L57) and [Configure](https://github.com/dotnet/aspnetcore/blob/a7ca111caeb3f1a66a65ca802a3b7d18617b2e80/src/Hosting/Hosting/src/WebHostBuilderExtensions.cs#L90) will try to pass through their calls to any IWebHostBuilder implementing ISupportsStartup giving us the opportunity to throw.

Currently only GenericWebHostBuilder implements ISupportsStartup and it lives in the same assembly so it can be internal. ConfigureWebHostBuilder is in another assembly that depends on Hosting, so it needs the interface to be public to implement it.

## Alternative considered
1.
    ```diff
    namespace Microsoft.AspNetCore.Hosting
    {
    +    public interface IRejectStartup
    +    {
    +        void ThrowException();
    +    }
    ```

    With this design, IRejectStartup is not just a marker interface, but it also throws an Exception via ThrowException() which would be called by the unsupported extension methods allowing any implementers of the interface to provide more detailed error messages. I'm not sure this is necessary, and if we wanted to support throwing different exceptions from Configure and UseStartup, the interface becomes more complicated.

2.
    ```diff
    namespace Microsoft.AspNetCore.Hosting
    {
    +    public interface ISupportsStartup
    +    {
    +        IWebHostBuilder Configure(Action<WebHostBuilderContext, IApplicationBuilder> configure);
    +        IWebHostBuilder UseStartup([DynamicallyAccessedMembers(StartupLinkerOptions.Accessibility)] Type startupType);
    +        IWebHostBuilder UseStartup<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]TStartup>(Func<WebHostBuilderContext, TStartup> startupFactory);
    +    }
    ```

     This feels like it exposes more than we have to and doesn't allow us to throw from ConfigureWebHost().

3. If ConfigureWebHostBuilder continued not to implement ISupportsStartup, we could look for additions of IStartup services to the [WebApplicationServiceCollection](https://github.com/dotnet/aspnetcore/blob/a7ca111caeb3f1a66a65ca802a3b7d18617b2e80/src/DefaultBuilder/src/WebApplicationServiceCollection.cs) instead and throw if that ever happens. This feels a lot less straightforward than implementing ISupportsStartup however.

4.
    ```diff
    namespace Microsoft.AspNetCore.Hosting
    {
    +    /// <summary>
    +    /// A marker interface used to indicate the <see cref="IHostBuilder"/> or <see cref="IWebHostBuilder"/>
    +    /// does not support any of the following extension methods:
    +    /// * <see cref="WebHostBuilderExtensions.Configure(IWebHostBuilder,         Action{AspNetCore.Builder.IApplicationBuilder})"/>
    +    /// * <see cref="WebHostBuilderExtensions.UseStartup(IWebHostBuilder, Type)"/>
    +    /// * <see cref="GenericHostWebHostBuilderExtensions.ConfigureWebHost(IHostBuilder, Action{IWebHostBuilder})"/>
    +    /// </summary>
    +    public interface IRejectStartup
    +    {
    +    }
```

This allows us to check inside these extension methods whether they should throw to indicate they're not supported by the given IHostBuilder or IWebHostBuilder (namely WebApplicationBuilder.Host/WebHost) rather than effectively doing nothing.

### PR DESCRIPTION
**Pull request by halter73**
WebApplicationBuilder.WebHost.UseStartup/Configure no-op today. This makes it clear that they don't work.

This now also causes WebApplicationBuilder.Host.ConfigureWebHost/ConfigureWebHostDefaults to throw for similar reasons. 

Fixes #34824

### CODE REVIEW

File: src/DefaultBuilder/src/ConfigureWebHostBuilder.cs [hunk 1]
```diff
@@ -153,5 +154,20 @@ public IWebHostBuilder UseSetting(string key, string? value)
...
         }
+
+        IWebHostBuilder ISupportsStartup.Configure(Action<WebHostBuilderContext, IApplicationBuilder> configure)
+        {
+            throw new NotSupportedException($"Configure() is not supported by WebApplicationBuilder.WebHost. Use the WebApplication returned by WebApplicationBuilder.Build() instead.");
+        }
+
+        IWebHostBuilder ISupportsStartup.UseStartup(Type startupType)
+        {
+            throw new NotSupportedException($"UseStartup() is not supported by WebApplicationBuilder.WebHost. Use the WebApplication returned by WebApplicationBuilder.Build() instead.");
```
Comment (halter73) [inline src/DefaultBuilder/src/ConfigureWebHostBuilder.cs:165]: I think these messages probably need work. Someone could also be interested in modifying services with the startup class, and for that they want to use `WebApplicationBuilder.Services` instead of `WebApplication`'s `IApplicationBuilder` implementation. I didn't want this to get too wordy though.

**Comment by davidfowl** [comment]
Also do the `builder.Host.ConfigureWebHost`

File: src/Hosting/Hosting/src/Startup/IRejectStartup.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+namespace Microsoft.AspNetCore.Hosting
+{
+    /// <summary>
+    /// A marker interface used to indicate the <see cref="IHostBuilder"/> or <see cref="IWebHostBuilder"/>
+    /// does not support any of the following extension methods:
+    /// * <see cref="WebHostBuilderExtensions.Configure(IWebHostBuilder, Action{AspNetCore.Builder.IApplicationBuilder})"/>
+    /// * <see cref="WebHostBuilderExtensions.UseStartup(IWebHostBuilder, Type)"/>
+    /// * <see cref="GenericHostWebHostBuilderExtensions.ConfigureWebHost(IHostBuilder, Action{IWebHostBuilder})"/>
+    /// </summary>
+    public interface IRejectStartup
```
Comment (halter73) [inline src/Hosting/Hosting/src/Startup/IRejectStartup.cs:15]: This is a new approach I'm taking instead of making `ISupportsStartup` public since it's a smaller change and it also works for `WebApplicationBuilder.Host.ConfigureWebHost`.

Comment (davidfowl) [reply src/Hosting/Hosting/src/Startup/IRejectStartup.cs:15]: This name seems weird when since this is being used to detect the host builder as well

Comment (halter73) [reply src/Hosting/Hosting/src/Startup/IRejectStartup.cs:15]: I'm open to suggestions.

Comment (davidfowl) [reply src/Hosting/Hosting/src/Startup/IRejectStartup.cs:15]: It's weird because interfaces themselves are about added capabilities so usually marker interfaces mark the presence of capability not the negation of it. 

- IDisallowWebHostConfiguration
- IBlockWebHostConfiguration

File: src/Hosting/Hosting/src/Startup/IRejectStartup.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Hosting;
+
+namespace Microsoft.AspNetCore.Hosting
+{
+    /// <summary>
+    /// A marker interface used to indicate the <see cref="IHostBuilder"/> or <see cref="IWebHostBuilder"/>
+    /// does not support any of the following extension methods:
+    /// * <see cref="WebHostBuilderExtensions.Configure(IWebHostBuilder, Action{AspNetCore.Builder.IApplicationBuilder})"/>
```
Comment (adityamandaleeka) [inline src/Hosting/Hosting/src/Startup/IRejectStartup.cs:11]: Need to use <list> for this: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#list

File: src/Hosting/Hosting/src/Startup/IRejectStartup.cs (hunk 2 above)
Comment (adityamandaleeka) [reply src/Hosting/Hosting/src/Startup/IRejectStartup.cs:15]: `ISupportConfigureWebHost`

**Comment by halter73** [comment]
⬆️📅

File: src/Hosting/Hosting/src/GenericHost/ISupportsConfigureWebHost.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Hosting;
+
+namespace Microsoft.AspNetCore.Hosting
```
Comment (davidfowl) [inline src/Hosting/Hosting/src/GenericHost/ISupportsConfigureWebHost.cs:6]:
```suggestion
namespace Microsoft.AspNetCore.Hosting.Infrastructure
```

File: src/Hosting/Hosting/src/GenericHost/ISupportsStartup.cs [hunk 5]
```diff
@@ -8,10 +8,33 @@
 
 namespace Microsoft.AspNetCore.Hosting
```
Comment (davidfowl) [inline src/Hosting/Hosting/src/GenericHost/ISupportsStartup.cs:9]:
```suggestion
namespace Microsoft.AspNetCore.Hosting.Infrastructure
```

**Comment by davidfowl** [comment]
Change the interface namespace.

**Review by davidfowl — APPROVED** [review]
